### PR TITLE
feat: upgrade storage crate to arrow and parquet offcial impl

### PR DESCRIPTION
**Pull request by v0y4g3r**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?
- Replace parquet reader and writer with `parquet` official crate
- Replace arrow ipc reader and writer with `arrow` official crate
- Fix all compile errors and clippy warnings in storage crate

## Checklist

- [X]  I have written the necessary rustdoc comments.
- [X]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
#555

### CODE REVIEW

File: src/storage/Cargo.toml [hunk 1]
```diff
@@ -7,6 +7,7 @@ license = "Apache-2.0"
 [dependencies]
 arc-swap = "1.0"
 async-stream = "0.3"
+async-compat = "0.2"
```
Comment (waynexia) [inline src/storage/Cargo.toml:10]:
```suggestion
async-compat = "0.2"
async-stream = "0.3"
```
😋 nit-picking

File: src/storage/src/memtable/btree.rs [hunk 2]
```diff
@@ -428,6 +426,7 @@ impl<'a> RowsProvider for &'a [&RowValue] {
     }
 }
 
+use datatypes::data_type::DataType;
```
Comment (waynexia) [inline src/storage/src/memtable/btree.rs:429]: maybe move this to file header?

File: src/storage/src/write_batch.rs [hunk 3]
```diff
@@ -500,14 +502,16 @@ impl<'a> IntoIterator for &'a WriteBatch {
     }
 }
 
+// TODO(hl): remove this allow
+#[allow(unused)]
```
Comment (evenyag) [inline src/storage/src/write_batch.rs:506]: Could we remove this now?

File: src/storage/src/write_batch.rs [hunk 4]
```diff
@@ -592,19 +594,14 @@ pub mod codec {
 
         fn decode(&self, src: &[u8]) -> Result<WriteBatch> {
             let mut reader = Cursor::new(src);
-            let metadata = read::read_stream_metadata(&mut reader).context(DecodeArrowSnafu)?;
-            let mut reader = StreamReader::new(reader, metadata);
-            let arrow_schema = reader.metadata().schema.clone();
-
+            // let metadata = read::read_stream_metadata(&mut reader).context(DecodeArrowSnafu)?;
```
Comment (evenyag) [inline src/storage/src/write_batch.rs:597]:
```suggestion
```

File: src/storage/src/sst/parquet.rs [hunk 5]
```diff
@@ -220,61 +135,51 @@ impl<'a> ParquetReader<'a> {
...
+        );
 
-        let projected_fields = adapter.fields_to_read();
-        let chunk_stream = try_stream!({
-            for (idx, valid) in pruned_row_groups.iter().enumerate() {
-                if !valid {
-                    debug!("Pruned {} row groups", idx);
-                    continue;
-                }
+        let projection = match self.projected_schema.projection() {
```
[the remark this answers is not in the document]
Comment (evenyag) [reply src/storage/src/sst/parquet.rs:162]: I think you could use `ReadAdapter::is_source_needed` to build the `ProjectionMask`, as how `ReadAdapter::fields_to_read()` does.

**Review by evenyag — COMMENTED** [review]
I'm afraid that the change to the `ParquetReader` is incorrect and may cause unexpected output.

File: src/storage/src/memtable/tests.rs [hunk 6]
```diff
@@ -576,22 +579,18 @@ fn test_memtable_projection() {
...
-        let v0 = UInt64Array::from_slice(&[10, 11, 12]);
-        let sequences = UInt64Array::from_slice(&[9, 9, 9]);
-        let op_types = UInt8Array::from_slice(&[0, 0, 0]);
-
-        assert_eq!(k0, &*batch.column(0).to_arrow_array());
-        assert_eq!(k1, &*batch.column(1).to_arrow_array());
-        assert_eq!(v0, &*batch.column(2).to_arrow_array());
-        assert_eq!(sequences, &*batch.column(3).to_arrow_array());
-        assert_eq!(op_types, &*batch.column(4).to_arrow_array());
+        let k0 = Arc::new(TimestampMillisecondVector::from_slice(vec![
```
Comment (evenyag) [inline src/storage/src/memtable/tests.rs:582]:
```suggestion
        let k0 = Arc::new(TimestampMillisecondVector::from_slice(&[
```

File: src/storage/src/metadata.rs [hunk 7]
```diff
@@ -670,7 +670,7 @@ struct ColumnsMetadataBuilder {
 }
 
 impl ColumnsMetadataBuilder {
-    fn row_key(&mut self, key: RowKeyDescriptor) -> Result<&mut Self> {
+    pub fn row_key(&mut self, key: RowKeyDescriptor) -> Result<&mut Self> {
```
Comment (evenyag) [inline src/storage/src/metadata.rs:673]:
```suggestion
   fn row_key(&mut self, key: RowKeyDescriptor) -> Result<&mut Self> {
```

File: src/storage/src/metadata.rs [hunk 8]
```diff
@@ -695,7 +695,7 @@ impl ColumnsMetadataBuilder {
         self.push_value_column(consts::KEY_CF_ID, desc)
     }
 
-    fn push_value_column(
+    pub fn push_value_column(
```
Comment (evenyag) [inline src/storage/src/metadata.rs:698]:
```suggestion
    fn push_value_column(
```

File: src/storage/src/metadata.rs [hunk 9]
```diff
@@ -734,7 +734,7 @@ impl ColumnsMetadataBuilder {
         Ok(self)
     }
 
-    fn build(mut self) -> Result<ColumnsMetadata> {
+    pub fn build(mut self) -> Result<ColumnsMetadata> {
```
Comment (evenyag) [inline src/storage/src/metadata.rs:737]:
```suggestion
    fn build(mut self) -> Result<ColumnsMetadata> {
```

File: src/storage/src/proto/write_batch.rs [hunk 10]
```diff
@@ -331,7 +376,12 @@ pub fn gen_put_data_vector(data_type: ConcreteDataType, column: Column) -> Resul
         ConcreteDataType::Float64(_) => gen_put_data_f64(column),
         ConcreteDataType::Binary(_) => gen_put_data_binary(column),
         ConcreteDataType::String(_) => gen_put_data_string(column),
-        ConcreteDataType::Timestamp(_) => gen_put_data_timestamp(column),
+        ConcreteDataType::Timestamp(t) => match t {
+            TimestampType::Second(_) => gen_put_data_ts_second(column),
+            TimestampType::Millisecond(_) => gen_put_data_ts_millisecond(column),
+            TimestampType::Microsecond(_) => gen_put_data_ts_microsecond(column),
+            TimestampType::Nanosecond(_) => gen_put_data_ts_nanosecond(column),
+        },
```
Comment (evenyag) [inline src/storage/src/proto/write_batch.rs:384]:
```rust
ConcreteDataType::Date(_) => gen_put_data_date(),
ConcreteDataType::DateTime(_) => gen_put_data_datetime(),
```

Comment (v0y4g3r) [reply src/storage/src/proto/write_batch.rs:385]: Also resolves `gen_column_xxx`
https://github.com/v0y4g3r/greptimedb/commit/a36f3996aaa0359da3ff285465a06b779880a9f4#diff-b62444418510c8b2cdfde57873bc8e1bfd927de3f1e9daab977cca83337b0ed2R348-R349

**Review by evenyag — APPROVED** [review]
LGTM

**Review by waynexia — APPROVED** [review]
LGTM

**Pull request merged**